src/outputs/history.cpp

```diff
@@ -49,68 +49,113 @@ void HistoryOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
+      md_base->Set(pm->block_list, pm);
+    } else if (md_base->NumBlocks() != pm->block_list.size()) {
+      PARTHENON_WARN(
+          "Resetting \"base\" MeshData to contain all blocks. This indicates that "
+          "the \"base\" MeshData container has been modified elsewhere. Double check "
+          "that the modification was intentional and is compatible with this reset.")
+      md_base->Set(pm->block_list, pm);
+    }
+
     for (const auto &hist_var : hist_vars) {
-      // Get "base" MeshData, which always exists but may not be populated yet
-      auto &md_base = pm->mesh_data.Get();
-      // Populated with all blocks
-      if (md_base->NumBlocks() == 0) {
-        md_base->Set(pm->block_list, pm);
-      } else if (md_base->NumBlocks() != pm->block_list.size()) {
-        PARTHENON_WARN(
-            "Resetting \"base\" MeshData to contain all blocks. This indicates that "
-            "the \"base\" MeshData container has been modified elsewhere. Double check "
-            "that the modification was intentional and is compatible with this reset.")
-        md_base->Set(pm->block_list, pm);
-      }
       auto result = hist_var.hst_fun(md_base.get());
-#ifdef MPI_PARALLEL
-      // need fence so that the result is ready prior to the MPI call
-      Kokkos::fence();
-      // apply separate chosen operations to each user-defined history output
-      MPI_Op usr_op;
-      switch (hist_var.hst_op) {
-      case UserHistoryOperation::sum:
-        usr_op = MPI_SUM;
-        break;
-      case UserHistoryOperation::max:
-        usr_op = MPI_MAX;
-        break;
-      case UserHistoryOperation::min:
-        usr_op = MPI_MIN;
-        break;
-      }
-      if (Globals::my_rank == 0) {
-        PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &result, 1, MPI_PARTHENON_REAL,
-                                       usr_op, 0, MPI_COMM_WORLD));
-      } else {
-        PARTHENON_MPI_CHECK(MPI_Reduce(&result, &result, 1, MPI_PARTHENON_REAL, usr_op, 0,
-                                       MPI_COMM_WORLD));
+      results[hist_var.hst_op].push_back(result);
+      labels[hist_var.hst_op].push_back(hist_var.label);
+    }
+
+    // Check if the package has enrolled vector history functions which are stored in the
+    // Params under the `hist_vec_param_key` name.
+    if (!params.hasKey(hist_vec_param_key)) {
+      continue;
+    }
+    const auto &hist_vecs = params.Get<HstVec_list>(hist_vec_param_key);
+    for (const auto &hist_vec : hist_vecs) {
+      auto result = hist_vec.hst_vec_fun(md_base.get());
+      for (int n = 0; n < result.size(); n++) {
+        std::string label = hist_vec.label + std::to_string(n);
+        results[hist_vec.hst_op].push_back(result[n]);
+        labels[hist_vec.hst_op].push_back(label);
       }
-#endif
+    }
+  }
 
-      all_results.emplace_back(result);
-      all_labels.emplace_back(hist_var.label);
+  // Need fence so result is ready prior to MPI call
+  Kokkos::fence();
```

Comment:
Should the fence be inside the `#ifdef MPI_PARALLEL`?

Reply:
You're right, fixed, thanks